File: back-kotlin/src/main/kotlin/me/nicolas/stravastats/domain/business/ActivityEffort.kt
package me.nicolas.stravastats.domain.business

import me.nicolas.stravastats.domain.utils.formatSeconds
import java.util.*


/**
 * An effort within an activity.
 */
class ActivityEffort(
    val distance: Double,
    val seconds: Int,
    val deltaAltitude: Double,
    val idxStart: Int,
    val idxEnd: Int,
    val averagePower: Int? = null,
    val label: String,
    val activityShort: ActivityShort
) {
    fun getFormattedSpeedWithUnits(): String {
        val speed = getFormatedSpeed()
        return if (activityShort.type in runActivities) {
            "${speed}/km"
        } else {
            "$speed km/h"
        }
    }

    fun getFormatedSpeed(): String {
        return if (activityShort.type in runActivities) {
            (seconds * 1000 / distance).formatSeconds()
        } else {
            "%.02f".format(Locale.ENGLISH, distance / seconds * 3600 / 1000)
        }
    }

    fun getMSSpeed(): Double {
        return distance / seconds
    }

    fun getFormattedGradientWithUnit() = "${this.getFormattedGradient()} %"

    fun getFormattedPower() = if (this.averagePower != null) "${this.averagePower} W" else ""

    fun getPower() = this.averagePower

    fun getFormattedGradient() = "%.02f".format(Locale.ENGLISH, getGradient())

    fun getGradient() =100 * deltaAltitude / distance

    fun getDescription() = "${this.label}:" +
            "<ul>" +
            "<li>Distance : %.1f km</li>".format(distance / 1000) +
            "<li>Time : ${seconds.formatSeconds()}</li>" +
            "<li>Speed : ${getFormattedSpeedWithUnits()}</li>" +
            "<li>Gradient: ${getFormattedGradient()}%</li>" +
            "<li>Power: ${if (averagePower != null) getFormattedPower() else "Not available"}</li>" +
            "</ul>"
}

File: back-go/domain/business/activityeffort.go
package business

import (
	"fmt"
	"strconv"
	"time"
)

type ActivityEffort struct {
	Distance      float64
	Seconds       int
	DeltaAltitude float64
	IdxStart      int
	IdxEnd        int
	AveragePower  *float64
	Label         string
	ActivityShort ActivityShort
}

func (ae ActivityEffort) GetFormattedSpeed() string {
	speed := ae.GetSpeed()
	if ae.ActivityShort.Type == Run {
		return fmt.Sprintf("%s/km", speed)
	}
	return fmt.Sprintf("%s km/h", speed)
}

func (ae ActivityEffort) GetSpeed() string {
	if ae.ActivityShort.Type == Run {
		return formatSeconds(ae.Seconds * 1000 / int(ae.Distance))
	}
	return fmt.Sprintf("%.02f", ae.Distance/float64(ae.Seconds)*3600/1000)
}

func (ae ActivityEffort) GetMSSpeed() float64 {
	speed := ae.Distance / float64(ae.Seconds)
	formattedSpeed, _ := strconv.ParseFloat(fmt.Sprintf("%.2f", speed), 64)
	return formattedSpeed
}

func (ae ActivityEffort) GetFormattedGradient() string {
	return fmt.Sprintf("%.02f%%", ae.GetGradient())
}

func (ae ActivityEffort) GetFormattedPower() string {
	if ae.AveragePower != nil {
		return fmt.Sprintf(`%d W`, int(*ae.AveragePower))
	}
	return "Not available"
}

func (ae ActivityEffort) GetGradient() float64 {
	return 100 * ae.DeltaAltitude / ae.Distance
}

func (ae ActivityEffort) GetDescription() string {
	return fmt.Sprintf("%s:<ul><li>Distance : %.1f km</li><li>Time : %s</li><li>Speed : %s</li><li>Gradient: %.02f%%</li><li>Power: %s</li></ul>",
		ae.Label,
		ae.Distance/1000,
		formatSeconds(ae.Seconds),
		ae.GetFormattedSpeed(),
		ae.GetGradient(),
		ae.GetFormattedPower())
}

func formatSeconds(seconds int) string {
	return time.Duration(seconds * int(time.Second)).String()
}
